File: ux/src/api/types/interchange.ts
/**
 * Interchange Type Definitions
 *
 * Type definitions for import/export operations (interchange)
 * Generated types are imported from the OpenAPI spec where available
 */

import type { components } from "@/api/client/types";

/**
 * SerializationScope - Scope parameters for serialization operations.
 * Describes what should be exported or imported.
 */
export type SerializationScope =
  components["schemas"]["SerializationScopeResponse"];

/**
 * SerializationScopeResponse - Describes what was serialized.
 */
export type SerializationScopeResponse =
  components["schemas"]["SerializationScopeResponse"];

/**
 * ResolutionRecord - Resolution applied to a conflict during import.
 */
export type ResolutionRecord =
  components["schemas"]["ResolutionRecordResponse"];

/**
 * ChangeEvent - Change event associated with an import run.
 */
export type ChangeEvent =
  components["schemas"]["InterchangeChangeEventResponse"];

/**
 * ChangeEventResponse - Response containing change event data.
 */
export type ChangeEventResponse =
  components["schemas"]["InterchangeChangeEventResponse"];

/**
 * ImportRunResponse - Response containing import run data.
 */
export type ImportRunResponse = components["schemas"]["ImportRunResponse"];

/**
 * ImportRunListResponse - Paginated list of import runs.
 * Uses generic ListResponse pattern.
 */
export type ImportRunListResponse =
  components["schemas"]["ListResponse_ImportRunResponse_"];

/**
 * ChangeEventListResponse - Paginated list of change events.
 * Uses generic ListResponse pattern.
 */
export type ChangeEventListResponse =
  components["schemas"]["ListResponse_InterchangeChangeEventResponse_"];

/**
 * Resolution strategy for import conflicts.
 */
export type ResolutionKind =
  | "skip"
  | "overwrite"
  | "merge"
  | "rename"
  | "abort";

/**
 * Incoming entity from import file.
 *
 * Shape varies by serialization format (SKOS, OWL, GraphML),
 * but the following fields are always extracted and available:
 * - id: Unique identifier for deduplication and keying
 * - title: Display label
 *
 * Additional fields (description, properties, etc.) depend on source format.
 */
export interface IncomingEntity {
  id: string;
  title: string;
  [key: string]: unknown;
}

/**
 * Conflict detected during import dry-run.
 */
export interface ImportConflict {
  match_kind: "external_reference" | "uuid" | "title";
  incoming: IncomingEntity;
  existing?: string;
  default_resolution: ResolutionKind;
  available_resolutions: ResolutionKind[];
}

/**
 * Result of an import dry-run, describing what would be imported.
 */
export interface ImportPlanResponse {
  conflicts: ImportConflict[];
  new_entity_count: number;
  import_run_id?: string;
  warnings: string[];
  source_hash?: string;
  scope?: SerializationScope;
}

/**
 * Status of an import run.
 */
export type ImportRunStatus =
  | "pending"
  | "committed"
  | "failed"
  | "rolled_back";

/**
 * Result of committing an import run.
 */
export type ImportRunCommitResponse = ImportRunResponse;

/**
 * Import run with change events.
 */
export type ImportRun = ImportRunResponse;

/**
 * Query parameters for listing import runs.
 */
export interface ImportRunListParams {
  offset?: number;
  limit?: number;
  status?: ImportRunStatus;
}

/**
 * Query parameters for listing change events.
 */
export interface ChangeEventListParams {
  offset?: number;
  limit?: number;
  entity_type?: string;
  change_type?: string;
}

File: local-server/adapters/web/schemas/interchange.py
"""
Pydantic schemas for the Data Interchange bounded context.

Request schemas:
- ExportRequest: For export operations
- SerializationScopeRequest: Describes what to export

Response schemas:
- SerializationScopeResponse: Describes what was serialized
- ImportConflictResponse: Conflict details
- ImportPlanResponse: Result of import dry-run
- ResolutionRecordResponse: Applied resolution
- ImportRunResponse: Import run details
- InterchangeChangeEventResponse: Change event details
"""

from datetime import datetime
from typing import Any, List, Literal, Optional

from pydantic import BaseModel, ConfigDict, Field

from domain.interchange.value_objects import ChangeOperation

# ==================== Serialization Scope Schemas ====================


class SerializationScopeRequest(BaseModel):
    """Request to specify what to export."""

    scope_type: Literal["whole_graph", "taxonomy", "scheme", "entity_set"] = Field(
        ..., description="Scope type: whole_graph, taxonomy, scheme, or entity_set"
    )
    taxonomy_id: Optional[str] = Field(
        None, description="Taxonomy ID for taxonomy scope"
    )
    scheme_id: Optional[str] = Field(None, description="Scheme ID for scheme scope")
    include_descendants: bool = Field(
        False, description="Include descendants for scheme scope"
    )
    entity_ids: Optional[List[str]] = Field(
        None, description="Entity IDs for entity_set scope"
    )


class SerializationScopeResponse(BaseModel):
    """Response describing what was serialized."""

    scope_type: Literal["whole_graph", "taxonomy", "scheme", "entity_set"]
    taxonomy_id: Optional[str] = None
    scheme_id: Optional[str] = None
    include_descendants: bool = False
    entity_ids: Optional[List[str]] = None


# ==================== Export Schemas ====================


class ExportRequest(BaseModel):
    """Request to export ontology data."""

    format: Literal["skos", "owl", "graphml"] = Field(
        ..., description="Export format: skos, owl, or graphml"
    )
    scope: SerializationScopeRequest = Field(..., description="What to export")
    split_mode: bool = Field(
        False,
        description=(
            "For OWL format: if true, export only TBox (schema) without ABox"
            " (individuals)"
        ),
    )


# ==================== Import Conflict and Plan Schemas ====================


class ImportConflictResponse(BaseModel):
    """Represents a conflict detected during import."""

    match_kind: Literal["external_reference", "uuid", "title"] = Field(
        ..., description="Type of match: external_reference, uuid, or title"
    )
    incoming: dict[str, Any] = Field(..., description="Incoming entity data")
    existing: Optional[str] = Field(None, description="Reference to existing entity")
    default_resolution: Optional[
        Literal["skip", "overwrite", "merge", "rename", "abort"]
    ] = Field(
        None, description="Default resolution strategy (None means user must choose)"
    )
    available_resolutions: List[
        Literal["skip", "overwrite", "merge", "rename", "abort"]
    ] = Field(..., description="Available resolutions")


class ResolutionRecordResponse(BaseModel):
    """Record of a resolution applied during import."""

    match_kind: Literal["external_reference", "uuid", "title"] = Field(
        ..., description="Type of match that was resolved"
    )
    entity_id: str = Field(..., description="Entity ID involved")
    resolution_chosen: Literal["skip", "overwrite", "merge", "rename", "abort"] = Field(
        ..., description="Resolution applied"
    )


class ImportPlanResponse(BaseModel):
    """Result of an import dry-run, describing what would be imported."""

    conflicts: List[ImportConflictResponse] = Field(
        default_factory=list, description="Detected conflicts"
    )
    new_entity_count: int = Field(
        ..., description="Number of new entities to be created"
    )
    import_run_id: Optional[str] = Field(None, description="Prospective import run ID")
    warnings: List[str] = Field(default_factory=list, description="Warning messages")
    source_hash: Optional[str] = Field(
        None, description="SHA256 hash of imported bytes"
    )
    scope: Optional[SerializationScopeResponse] = Field(
        None, description="Import scope"
    )


# ==================== Import Run Schemas ====================


class ImportRunResponse(BaseModel):
    """Response containing import run data."""

    model_config = ConfigDict(from_attributes=True)

    id: str = Field(..., description="Unique identifier")
    created_at: datetime = Field(..., description="Creation timestamp")
    created_by: Optional[str] = Field(None, description="User who initiated the import")
    format: Literal["skos", "owl", "graphml"] = Field(
        ..., description="Format of imported file"
    )
    source_uri: Optional[str] = Field(None, description="URI or filename of source")
    source_hash: str = Field(..., description="SHA256 hash of imported bytes")
    scope: SerializationScopeResponse = Field(..., description="Import scope")
    resolutions: List[ResolutionRecordResponse] = Field(
        default_factory=list, description="Applied conflict resolutions"
    )
    affected_entity_ids: List[str] = Field(
        default_factory=list, description="Entities affected by import"
    )
    warnings: List[str] = Field(
        default_factory=list, description="Warning messages from import operation"
    )
    status: Literal["pending", "committed", "failed", "rolled_back"] = Field(
        ..., description="Current status: pending, committed, failed, or rolled_back"
    )


# ==================== Change Event Schemas ====================


class InterchangeChangeEventResponse(BaseModel):
    """Response containing interchange change event data."""

    model_config = ConfigDict(from_attributes=True)

    id: str = Field(..., description="Unique identifier")
    timestamp: datetime = Field(..., description="When the change occurred")
    entity_id: str = Field(..., description="Entity that changed")
    entity_type: str = Field(..., description="Type of entity")
    operation: ChangeOperation = Field(..., description="Operation performed")
    new_state: Optional[dict[str, Any]] = Field(
        None, description="New state after change"
    )
    previous_state: Optional[dict[str, Any]] = Field(
        None, description="Previous state before change"
    )
